third_party/move/move-compiler/src/typing/expand.rs

```diff
@@ -57,43 +57,84 @@ fn types(context: &mut Context, ss: &mut Vec<Type>) {
     }
 }
 
-pub fn type_(context: &mut Context, ty: &mut Type) {
+// type_ for expanding the `i`-th type param of struct `ty_name`
+fn type_struct_ty_param(context: &mut Context, ty: &mut Type, i: usize, ty_name: &TypeName_) {
+    match ty_name {
+        TypeName_::ModuleType(mod_id, struct_name) => {
+            let param_name = &context.struct_definition(mod_id, struct_name).type_parameters[i].param.user_specified_name.value;
+            let msg = format!("Cannot infer the type parameter {param_name} for generic struct {ty_name}. Try providing a type parameter.");
+            type_msg_(context, ty, &msg);
+        },
+        _ => type_(context, ty),
+    }
+}
+
+fn type_fun_ty_param(context: &mut Context, ty: &mut Type, i: usize, m: &ModuleIdent, n: &FunctionName) {
+   let param_name = context.function_info(m, n).signature.type_parameters[i].user_specified_name;
+   type_msg_(context, ty, &format!("Cannot infer the type parameter {param_name} for generic function {m}::{n}. Try providing a type parameter."));
+}
+
+pub fn type_msg_(context: &mut Context, ty: &mut Type, msg_uninferred: &str) {
```

Comment:
This name isn't not self-explanatory.  Add a comment, perhaps something like:
```
/// Try to expand the type of ty, warning with msg_uninferred if type cannot be inferred.
```
HOWEVER, I think it would be better to change this method to `try_to_infer_type(context, ty): bool` which returns true on success, and have the message be created only if the it returns failure.  This would avoid creating a lot of unneeded warning messages, just in case they might be needed.

Reply:
I was halfway transitting, and realized this: there cannot be a `try_to_infer_type(context, ty) -> bool` that lets the caller choose what to do when type inference fails. Consider when we are expanding `S<X, Y>`, and we cannot infer `X`. The caller only gets a `false`, and doesn't know which parameter cannot be inferred, so no way to provide proper diagnostics.

Admittedly, `type_msg_` is a bad name, but I derived that from the original `type_`, which uses a general warning for uninferred types.

Hence I will maintain the current design, and add comments for `type_msg_`.